### ISSUE TITLE
validate over time 

### ISSUE THREAD

**Issue by mlindauer** [automl/SMAC3#319]
In our current smac-validate, we use the time steps 2^1,2^2,2^3,...

SMAC2 however uses cutoff/2^0, cutoff/2^1, cutoff/2^2, cutoff/2^3.
This has the big advantage that the end point of the trajectory is part of the validation.

Therefore, @shukon please change the time series in smac-validate accordingly.

**Comment by shukon** [comment]
I will do that with #316, just noting that validation with `time` always validates first and last incumbent anyway [source](https://github.com/automl/SMAC3/blob/development/smac/utils/validate.py#L355-L364)

```
 if mode == "time":
    configs.append(self.traj[0]["incumbent"])  # add first
    counter = 2^(-4)
    for entry in self.traj[:-1]:
        if (entry["wallclock_time"] >= counter and
                    entry["incumbent"] not in configs):
             configs.append(entry["incumbent"])
             counter *= 2
    configs.append(self.traj[-1]["incumbent"]) # add last
```

**Comment by shukon** [comment]
I assume, that cutoff in this context is the `wallclock-limit`.
I now implemented your proposal, but instead of the `wallclock_limit` specified in the scenario, the maximum time is used (`trajectory[-1]['wallclock_time']`). This is done in order to avoid problems when `wallclock_limit` is set to `inf`. [(commit)](https://github.com/automl/SMAC3/pull/323/commits/80b654d7963bcf6e8620b661f018e8e9bf960f21)
In addition, after looking into the code of SMAC2, the user can now choose between `wallclock_time` and `cpu_time`, instead of using the generic `time`-option. [(commit)](https://github.com/automl/SMAC3/pull/323/commits/e46550f461330436710623138230c1ae892dfff0)
@mlindauer Do you approve this or is there a disadvantage?

**Comment by mlindauer** [comment]
> I assume, that cutoff in this context is the wallclock-limit.

Yes.

> I now implemented your proposal, but instead of the wallclock_limit specified in the scenario, the maximum time is used (trajectory[-1]['wallclock_time']). This is done in order to avoid problems when wallclock_limit is set to inf.

That won't work. Since the last entry of the trajectory will have different time stamps, we won't have the same time series for all configurator runs; however, I would like that the time series is always the same for a given wallclock budget.

> In addition, after looking into the code of SMAC2, the user can now choose between wallclock_time and cpu_time, instead of using the generic time-option.

`cpu_time` in the sense of the time used only by the target algorithm? 
Actually, I don't know whether this is an important feature. However, since it is only using a different field of the trajectory, it shouldn't hurt to implement it, right?

**Comment by shukon** [comment]
> Since the last entry of the trajectory will have different time stamps, we won't have the same time series for all configurator runs; however, I would like that the time series is always the same for a given wallclock budget.

How about using wallclock-limit if its not inf and using maximum time otherwise?

**Comment by shukon** [comment]
@mlindauer now implemented the mentioned variation (limit, if available, else maxtime) in #323 . It makes perfect sense to use always the same timesteps, if possible.

**Comment by shukon** [comment]
fixed in #323
